### ISSUE TITLE
Impossible Mission II: Visual Glitch on Tandy

### ISSUE THREAD

**Issue by weirddan455** [dosbox-staging/dosbox-staging#3447]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### What Operating System are you using?

Linux x86_64

### If Other OS, please describe

Arch Linux

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

As the first level is loading in, there is an animation where the screen starts out as a narrow band in the center and then expands out vertically in both directions until it fills the screen completely.

This animation looks correct in v0.80.1 and incorrect on main.  This is a regression that I have bisected to commit dd8ae1423a9d496f78fdcf39798568aea6bbadca

Note there is a (mostly) unrelated divide by zero crash that must be worked around if checking out older commits.

### Steps to reproduce the behaviour.

Explain how to reproduce
1.  Put `machine = tandy` in your config
2. Run IM.COM or IM2_TAN .EXE (the former seems to auto-detect machine type and load the latter)
3. Skip past intro and "Load save game" screen (just press enter)
4. Observe the animation as the level loads in

### Download URL of affected game or software

https://www.abandonwaredos.com/abandonware-game.php?abandonware=Impossible+Mission+2&gid=1873

### Your configuration

```ini
[dosbox]
machine = tandy
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by weirddan455** [comment]
I think I've narrowed the problem down to pixel aspect ratio.  It's changing the resolution in small increments from around 160x10 to 160x200.  It looks like we are taking that 160x100 image and stretching it across the entire screen when we should be drawing a narrow band in the center of the screen and leaving the top/bottom black.

I'm not going to dive deeper into this tonight but it looks like the simple calculation of `PAR = width / height` is not sufficient for this case.  Granted this is a strange edge case and I'm not even sure it's possible to do this on VGA.  We may need some Tandy specific hack.

**Comment by johnnovak** [comment]
> I'm not going to dive deeper into this tonight but it looks like the simple calculation of `PAR = width / height` is not sufficient for this case. Granted this is a strange edge case and I'm not even sure it's possible to do this on VGA. We may need some Tandy specific hack.

Ah, from memory I may have simplified the PAR calculation for the Tandy, thinking weird custom modes are a VGA-only thing anyway. So restoring the pre-refactor behaviour for the affected Tandy path only hould restore the behaviour.

**Comment by weirddan455** [comment]
>  Ah, from memory I may have simplified the PAR calculation for the Tandy, thinking weird custom modes are a VGA-only thing anyway. So restoring the pre-refactor behaviour for the affected Tandy path only would restore the behaviour.

You actually added a bunch of complexity to it that I'm struggling to figure out.  The PAR for Tandy was hard-coded to 5/6.  Applying just that change to https://github.com/dosbox-staging/dosbox-staging/commit/dd8ae1423a9d496f78fdcf39798568aea6bbadca fixes it but on main even more complicated aspect ratio calculations got added so now it distorts the entire image to a very wide resolution if I try to do that.

**Comment by dani-jozsef** [comment]
As far as I understand, unlike VGA monitors, CGA and Tandy ones had a fixed hardware resolution, so I'm wondering if aspect ratio should be calculated from the known hardware limitations as opposed to logical screen resolution?

**Comment by johnnovak** [comment]
> You actually added a bunch of complexity to it that I'm struggling to figure out. The PAR for Tandy was hard-coded to 5/6. Applying just that change to [dd8ae14](https://github.com/dosbox-staging/dosbox-staging/commit/dd8ae1423a9d496f78fdcf39798568aea6bbadca) fixes it but on main even more complicated aspect ratio calculations got added so now it distorts the entire image to a very wide resolution if I try to do that.

Well, it's a complex topic/functionality :smile:

Yeah now I remember, that hardcoding seemed wrong to me. Taking a step back, we don't know for certain that the "zooming in" effect is what happens on real hardware, but I guess we csn simply restore previous behaviour.

I can do that sometime if you're stuck (I think you're not accounting for the optional pixel doubling when setting the hardcoded PAR).

**Comment by weirddan455** [comment]
Yeah it was the pixel doubling not being accounted for.  I think I can work out a fix for this.  I suppose we would need real hardware to know 100% what the correct behavior is but the current behavior just looks very wrong compared to v0.80.1.

This effect is not just being used when the level first gets loaded in.  It's also used for every screen change (just walk the character to the edge of the screen to reproduce).

Additionally, I'm noticing a flickering when this effect happens that shouldn't be there, even the after the PAR issue is fixed.

**Comment by johnnovak** [comment]
> Yeah it was the pixel doubling not being accounted for. I think I can work out a fix for this. I suppose we would need real hardware to know 100% what the correct behavior is but the current behavior just looks very wrong compared to v0.80.1.
> 
> This effect is not just being used when the level first gets loaded in. It's also used for every screen change (just walk the character to the edge of the screen to reproduce).

Yeah, please PR a fix. It's good that someone else becomes a bit more familiar with that VGA code too. Indeed, the game must use this as a cool transition effect then (a "hardware effect") if this happens on every screen change.

Make sure to test your fix with both `crt-auto` (does pixel doubling), and `crt-auto-arcade` and `sharp` (no pixel doubling).

> Additionally, I'm noticing a flickering when this effect happens that shouldn't be there, even the after the PAR issue is fixed.

Might be because of this:

```
commit b68b66f07831c91f197acf0b06f847c28c9a7946
Author: kcgen <kcgen@users.noreply.github.com>
Date:   Wed May 24 10:26:22 2023 -0700

    Resize VGA within half the current frame-time (#2510)
```

That's a questionable change; kcgen did it to fix Jazz Jackrabbit, from memory, but I don't think real monitors can react to resolution changes that quickly (in single-digit milliseconds). So we might need to revert it.

**Comment by weirddan455** [comment]
>  As far as I understand, unlike VGA monitors, CGA and Tandy ones had a fixed hardware resolution vertically, so I'm wondering if aspect ratio should be calculated from the known hardware limitations as opposed to logical screen resolution?

Scratch what I said before.  I was looking at a technical reference manual for Tandy 100 and it looks like it had a typo implying "medium resolution" was 200x320.  Really it is 320x200 I believe (King's Quest 4 uses this mode).  This means we should be able to use a hard-coded 200 height when calculating PAR.

>  That's a questionable change; kcgen did it to fix Jazz Jackrabbit, from memory, but I don't think real monitors can react to resolution changes that quickly (in single-digit milliseconds). So we might need to revert it.

Yep, this was the thing that caused the flickering.  It was a fix for Earthworm Jim.  I really don't want to deal with more regressions so I just reverted to a 50ms delay for Tandy only.

PR coming up now.

**Comment by johnnovak** [comment]
> Yep, this was the thing that caused the flickering. It was a fix for Earthworm Jim. I really don't want to deal with more regressions so I just reverted to a 50ms delay for Tandy only.

That's cool, just put an explanatory comment there.

**Comment by dani-jozsef** [comment]
> Yep, this was the thing that caused the flickering. It was a fix for Earthworm Jim. I really don't want to deal with more regressions so I just reverted to a 50ms delay for Tandy only.

Since this was a workaround for a specific VGA game, wouldn't it be cleaner to have the behavior 'for VGA only'?

**Comment by weirddan455** [comment]
>  Since this was a workaround for a specific VGA game, wouldn't it be cleaner to have the behavior 'for VGA only'?

Maybe but these both seem to be edge cases.  Since we're planning on releasing a bugfix 0.81.1 patch in order to address some regressions, I'd rather these fixes be as conservative/targeted as possible to avoid causing even more regressions.

**Comment by johnnovak** [comment]
> > Since this was a workaround for a specific VGA game, wouldn't it be cleaner to have the behavior 'for VGA only'?
> 
> Maybe but these both seem to be edge cases. Since we're planning on releasing a bugfix 0.81.1 patch in order to address some regressions, I'd rather these fixes be as conservative/targeted as possible to avoid causing even more regressions.

Agreed.
